### ISSUE TITLE
CF: incorrect errors for ragged array structure

### ISSUE THREAD

**Issue by mhidas** [ioos/compliance-checker#845]
We have some files structured according to the [CF indexed ragged array representation](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_indexed_ragged_array_representation). The basic structure looks like this:
```
netcdf indexed_ragged_array {
dimensions:
        OBSERVATION = 12 ;
        INSTRUMENT = 3 ;
variables:
        double TIME(OBSERVATION) ;
                TIME:axis = "T" ;
        float TEMP(OBSERVATION) ;
                TEMP:coordinates = "TIME LATITUDE LONGITUDE NOMINAL_DEPTH" ;
        int instrument_index(OBSERVATION) ;
                instrument_index:instance_dimension = "INSTRUMENT" ;
                instrument_index:long_name = "which instrument this obs is for" ;
        double LATITUDE(INSTRUMENT) ;
                LATITUDE:axis = "Y" ;
        double LONGITUDE(INSTRUMENT) ;
                LONGITUDE:axis = "X" ;
        float NOMINAL_DEPTH(INSTRUMENT) ;
                NOMINAL_DEPTH:axis = "Z" ;
}
```
As far as I can tell, this is a correct implementation of this valid CF representation.
When I run the CF checker on it, I get the following errors and warnings:
```
                                    Errors                                     
--------------------------------------------------------------------------------
§5 Coordinate Systems
* dimensions for auxiliary coordinate variable LATITUDE (INSTRUMENT) are not a subset of dimensions for variable TEMP (OBSERVATION)
* dimensions for auxiliary coordinate variable LONGITUDE (INSTRUMENT) are not a subset of dimensions for variable TEMP (OBSERVATION)
* dimensions for auxiliary coordinate variable NOMINAL_DEPTH (INSTRUMENT) are not a subset of dimensions for variable TEMP (OBSERVATION)

§9.1 Features and feature types
* Unidentifiable feature for variable TEMP
* Unidentifiable feature for variable instrument_index


                                    Warnings                                    
--------------------------------------------------------------------------------
§3.1 Units
* units (None) attribute of 'instrument_index' must be a string compatible with UDUNITS
```
The `Unidentifiable feature` seems to be a known issue (#719) not specific to this structure.
The other two are checks that need to make an exception for this kind of file.

**Comment by mhidas** [comment]
Note that the indexed ragged array representation should be easy to identify, as stated in CF:
> In this representation, the file contains an index variable , which must be of type integer, and must have the sample dimension as its single dimension. The index variable contains the zero-based index of the feature to which each element belongs. This representation is identifiable by the presence of an attribute, instance_dimension , on the index variable, which names the dimension of the instance variables.

Also, I suspect the same issue would exist for the [contiguous ragged array representation](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_contiguous_ragged_array_representation), and similarly
> This representation and its count variable are identifiable by the presence of an attribute, sample_dimension , found on the count variable, which names the sample dimension being counted.

**Comment by daltonkell** [comment]
Thanks @mhidas for bringing this to the forefront. Coincidentally, I was testing data found on some ERDDAP servers just yesterday and realized that a similar result was happening with the data in the `.ncCF` (contiguous ragged array) and `.ncCFMA` (multidimensional array) representations.

This is a bit outside the scope of the current milestone, but I'm going to add it to the next one to ensure we don't lose track of it. I'd like to get to it ASAP, as many datasets use these formats to conserve space; they should not be marked as invalid for doing so.

**Comment by daltonkell** [comment]
Upon closer examination @mhidas, I think this issue actually references two pieces of the spec:
  1. Determining appropriate dimension structure for certain DSG auxiliary coordinate variables
  2. Identifying types of features within DSG

You aptly reference #719, which is addressed by #835 (and needs to be merged), which connects to (2) in the above list. The _other_ item, (1), is _not yet addressed_ by any issue as far as I can tell. One that comes close is https://github.com/ioos/compliance-checker/issues/284, a thread which you actually contributed to... and is four years old.

It seems that the checker correctly identified your dataset as one with the orthogonal timeseries representation, but something is amiss with the auxiliary coordinate variables. For this purpose, I am considering opening up a new issue (I know, issue hell) that pertains specifically to this which we can then assign to the new 4.4.0 milestone. I first have to understand what the problem is with auxiliary coordinate variables.

Once again, thanks for your continued involvement and diligence. Your perspective is greatly appreciated.

**Comment by mhidas** [comment]
Thanks @daltonkell . I'm glad you'll be able to look into this for the next release. Unfortunately I don't have much time to contribute to fixing this issue right now. 

> It seems that the checker correctly identified your dataset as one with the orthogonal timeseries representation, but something is amiss with the auxiliary coordinate variables.

Sorry, I think there's a bit of confusion here. If you're referring to the CDL file structure I described [above](https://github.com/ioos/compliance-checker/issues/845#issue-678303225), that is definitely **not** an "orthogonal timeseries representation", it is an "indexed ragged array".

This representation is identifiable by its index variable (`instrument_index` in my example), which has the attribute `instance_dimension`, identifying the dimension of the instance variables (i.e. referring to a particular feature within the file - see [definitions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#collections-instances-elements)). The dimension of index variable is the _sample dimension_, (which enumerates individual observations). In this representaion
* Geophysical variables should have the sample dimension as their only dimension (`OBSERVATION` in my example); and
* Auxiliary coordinate variables should have either the instance dimension as their only dimension (`INSTRUMENT`), or the sample dimension (as for the variable `TIME` in my example).

That's very different from the [orthogonal](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_orthogonal_multidimensional_array_representation) or [incomplete](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_incomplete_multidimensional_array_representation) multidimensional array representations, where the auxiliary coordinates need to have a subset of the geophysical variable's dimensions.

These different requirements would also be easy to implement, once the file is identified as an indexed ragged array file.

**Comment by mhidas** [comment]
I no longer get the incorrect errors for the ragged array files I have, so the recent work on feature type checks has fixed this issue too.
